### ISSUE TITLE
Packet losses with VALE

### ISSUE THREAD

**Issue by Linerd** [luigirizzo/netmap#441]
Hi,

I am seeing a bit under-performance of the VALE switch in my setup, I only ran the following commands:
sudo ./pkt-gen -i vale0:rx
sudo ./pkt-gen -i vale0:tx -f tx -R 900000 -l 1500 -n 10000000
The rate is not even 1Mpps yet, and the receiver side sees tens of thousands of packet loss.
I changed the packet len to default, and the same thing happened.
From previous threads this is due to VALE ports being non-blocking.
I tried connecting two pkt-gens with netmap patched veth ports and the packet rate increases substantially. But it's not ideal if VALE switch can't be used to interconnect different veth pairs (without significantly dropping the packet rate).

OS: 4.13.0-32-generic, Ubuntu 16.04

BTW, in the netmap-tutorial repo, is mymodule up to date with the current netmap? I ran into kernel errors when trying to insert mymodule_lin.ko.

Thanks!

**Comment by vmaffione** [comment]
VALE is not blocking, so you may have packet losses across two VALE ports independently from the packet rate. If the receiver for some reason does not drain fast enough, or if the traffic is too bursty and causes RX ring overflow, you will see packet drops.
It may be even CPU scheduler's fault, if it does not give enough CPU (and/or very often) to the receiver.
This is not a performance limitation of VALE, it is just a consequence of VALE design; OpenVSwitch and in-kernel Linux bridge have the same problem, as the switch is not blocking and overflow can happen on the TAP devices.

On my machine your tests show no packet drops (10000000 packets received). So it depends on how fast the machine is and how processes are scheduled.

**Comment by Linerd** [comment]
Thanks for the reply. However I was able to generate 20+Mpps with netmap-patched veth ports(without VALE), so I am not too certain if this is receiver's fault or CPU scheduler's fault.

BTW, how well do you expect (or maybe as you've tested) netmap to perform with parallel flows. Say the single flow packet rate I am able to generate is 30Mpps (with 64B packets), should I expect 1000 flows to each enjoy 30Kpps? (say I launch 1000 pairs of pkt-gens connected with veth, and make sure those pkt-gens have enough CPU resources).

**Comment by vmaffione** [comment]
Never tested something like that.
If you also use 1000 independent veth pairs and have enough CPUs, and memory is not a bottleneck, you ideally would have 30 Mpps for each flows. But this is just idealistic, as memory would be the bottleneck.

If you want to multiplex 1000 flows over a single veth pair you'd have to provide some form of external locking, which would kill performance before of lock contention. Or you could decouple clients from the demultiplexer (that runs in a thread) by using use a scheduling architecture like this http://info.iet.unipi.it/~luigi/papers/20160921-pspat.pdf

**Comment by Linerd** [comment]
Well I definitely would not expect 30Mpps per flow. Ideally I would like to
see 30kpps per flow, with independent veth pairs, all running on the same
core. I am just interested in knowing if the amount of resource can be
evenly distributed to different netmap applications without performance
degradation

**Comment by Linerd** [comment]
@vmaffione I did a bit of experiment today and found the following:

I modified the sample mswitch module to broadcast packets coming from port 0, and to send to port 0 coming from all other ports. And I created vale0 and vale1, which are connected with 10 netmap-bridge instances, while pkt-gen sends at 100Mbps from vale0:port0, and receives at vale1:port0.

**Normally one netmap-bridge forwarding at 100Mbps (1500B) consumes around 2% CPU, but when ten instances are running, each consumes more than 10%.** Receiver side pkt-gen sees 1Gbps throughput, which is 10X the sender throughput.
